src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -385,10 +387,206 @@ public static void EnsureSocketsAreInitialized()
+            }
+        }
+
+        private static unsafe IPAddress CreateIPv4Address(ReadOnlySpan<byte> socketAddress)
+        {
+            long address = (long)SocketAddressPal.GetIPv4Address(socketAddress) & 0x0FFFFFFFF;
+            return new IPAddress(address);
+        }
+
+        private static unsafe IPAddress CreateIPv6Address(ReadOnlySpan<byte> socketAddress)
+        {
+            Span<byte> address = stackalloc byte[IPAddressParserStatics.IPv6AddressBytes];
+            uint scope;
+            SocketAddressPal.GetIPv6Address(socketAddress, address, out scope);
+
+            return new IPAddress(address, (long)scope);
+        }
+
+        #region GetAddrInfoAsync Helper Classes
+
+        //
+        // Warning: If this ever ported to NETFX, AppDomain unloads needs to be handled
+        // to protect against AppDomainUnloadException if there are pending operations.
+        //
+
+        private sealed class GetAddrInfoExState
+        {
+            private DnsResolveAsyncResult _asyncResult;
+            private object _result;
+
+            public string HostName => _asyncResult.HostName;
+
+            public GetAddrInfoExState(DnsResolveAsyncResult asyncResult)
+            {
+                _asyncResult = asyncResult;
+            }
+
+            public void CompleteAsyncResult(object o)
+            {
+                // We don't want to expose the GetAddrInfoEx callback thread to user code.
+                // The callback occurs in a native windows thread pool.
+
+                _result = o;
+
+                Task.Factory.StartNew(s =>
+                {
+                    var self = (GetAddrInfoExState)s;
+                    self._asyncResult.InvokeCallback(self._result);
+                }, this, CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
+            }
+
+            public IntPtr CreateHandle()
+            {
+                GCHandle handle = GCHandle.Alloc(this, GCHandleType.Normal);
+                return GCHandle.ToIntPtr(handle);
+            }
+
+            public static GetAddrInfoExState FromHandle(IntPtr handle)
+            {
+                GCHandle gcHandle = GCHandle.FromIntPtr(handle);
+                var state = (GetAddrInfoExState)gcHandle.Target;
+                gcHandle.Free();
+
+                return state;
+            }
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        private unsafe struct GetAddrInfoExContext
+        {
+            private static readonly int Size = sizeof(GetAddrInfoExContext);
+
+            public NativeOverlapped Overlapped;
+            public AddressInfoEx* Result;
+            public IntPtr CancelHandle;
+            public IntPtr QueryStateHandle;
+
+            public static GetAddrInfoExContext* AllocateContext()
+            {
+                var context = (GetAddrInfoExContext*)Marshal.AllocHGlobal(Size);
+                *context = new GetAddrInfoExContext();
```

Comment:
Nit: might be slightly more clear if this were just `*context = default;`, but not a big deal.